internal/db/p2p_replicator.go

```diff
@@ -335,3 +344,377 @@ func (db *db) loadAndPublishReplicators(ctx context.Context) error {
 	}
 	return nil
 }
+
+// retryStatus is used to communicate if the retry was successful or not.
+type retryStatus struct {
+	PeerID  string
+	Success bool
+}
+
+// handleReplicatorRetries manages retries for failed replication attempts.
+func (db *db) handleReplicatorRetries(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case r := <-db.retryChan:
+			err := db.handleReplicatorFailure(ctx, r)
+			if err != nil {
+				log.ErrorContextE(ctx, "Failed to handle replicator failure", err)
+			}
+
+		case r := <-db.retryDone:
+			err := db.handleCompletedReplicatorRetry(ctx, r)
+			if err != nil {
+				log.ErrorContextE(ctx, "Failed to handle completed replicator retry", err)
+			}
+
+		case <-time.After(retryLoopInterval):
+			db.retryReplicators(ctx)
+		}
+	}
+}
+
+func (db *db) handleReplicatorFailure(ctx context.Context, r event.ReplicatorFailure) error {
+	err := db.updateReplicatorStatus(ctx, r.PeerID.String(), false)
+	if err != nil {
+		return err
+	}
+	err = db.createIfNotExistsReplicatorRetry(ctx, r.PeerID.String())
+	if err != nil {
+		return err
+	}
+	docIDKey := core.NewReplicatorRetryDocIDKey(r.PeerID.String(), r.DocID)
+	err = db.Peerstore().Put(ctx, docIDKey.ToDS(), []byte{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (db *db) handleCompletedReplicatorRetry(ctx context.Context, r retryStatus) error {
+	if r.Success {
+		done, err := db.deleteReplicatorRetryIfNoMoreDocs(ctx, r.PeerID)
+		if err != nil {
+			return err
+		}
+		if done {
+			err := db.updateReplicatorStatus(ctx, r.PeerID, true)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		err := db.setReplicatorNextRetry(ctx, r.PeerID)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// updateReplicatorStatus updates the status of a replicator in the peerstore.
+func (db *db) updateReplicatorStatus(ctx context.Context, peerID string, active bool) error {
+	key := core.NewReplicatorKey(peerID)
+	repBytes, err := db.Peerstore().Get(ctx, key.ToDS())
+	if err != nil {
+		return err
+	}
+	rep := client.Replicator{}
+	err = json.Unmarshal(repBytes, &rep)
+	if err != nil {
+		return err
+	}
+	switch active {
+	case true:
+		rep.Status = client.ReplicatorStatusActive
+		if rep.Status == client.ReplicatorStatusInactive {
+			rep.LastStatusChange = time.Time{}
+		}
+	case false:
+		rep.Status = client.ReplicatorStatusInactive
+		if rep.Status == client.ReplicatorStatusActive {
+			rep.LastStatusChange = time.Now()
+		}
+	}
+	b, err := json.Marshal(rep)
+	if err != nil {
+		return err
+	}
+	return db.Peerstore().Put(ctx, key.ToDS(), b)
+}
+
+type retryInfo struct {
+	NextRetry  time.Time
+	NumRetries int
+	Retrying   bool
+}
+
+func (db *db) createIfNotExistsReplicatorRetry(ctx context.Context, peerID string) error {
+	key := core.NewReplicatorRetryIDKey(peerID)
+	exists, err := db.Peerstore().Has(ctx, key.ToDS())
+	if err != nil {
+		return err
+	}
+	if !exists {
```

Comment:
suggestion: Inverting this if would remove a level of indentation (and complexity) from the bulk of this function:
```go
if exists {
  return
}
r := ...
```

Reply:
✅